.travis.yml

```diff
@@ -14,24 +14,147 @@ git:
-  - echo "Testing ssh to github.com" && ($GIT_SSH -T "git@github.com" || true)
+  - if [[ ! -d "${HOME}/mpich/build" ]] ; then mkdir -p "${HOME}/mpich/build" ; fi
+  - if [[ ! -d "${HOME}/usr" ]] ; then mkdir -p "${HOME}/usr/{bin,lib,share,include,lib64,libexec}" && export PATH="${HOME}/usr/bin:${PATH}"; fi
+  - |
+    if [[ -n "${encrypted_c48b4c75cf28_key}" ]] ; then # not a PR or someone else's repo
+      openssl aes-256-cbc -K $encrypted_c48b4c75cf28_key -iv $encrypted_c48b4c75cf28_iv -in id_rsa_Travis2GH.enc -out id_rsa_Travis2GH -d \
+      && chmod 600 id_rsa_Travis2GH \
+      && printf '#!/usr/bin/env bash\nssh -i %s/id_rsa_Travis2GH "$@"\n' "$(pwd)" > "${HOME}/bin/ssh_to_GH" \
+      && chmod +x "${HOME}/bin/ssh_to_GH" \
+      && export GIT_SSH="${HOME}/bin/ssh_to_GH" \
+      && echo "Testing ssh to github.com" && ($GIT_SSH -T "git@github.com" || true)
+    fi
+  - unset -v ${!encrypted_@} # unset secret encrypted variables
   - git --version
   - env | grep -iv 'token\|key\|encrypt' | sort # scrub tokens from output & sort
   - mount
+  - ls -l $(which gcc) && gcc --version # dump info about compilers
+  - ls -l $(which g++) && g++ --version
+  - if which gfortran; then ls -l $(which gfortran) && gfortran --version ; fi
+  - if ((RANDOM % 2 == 0)); then SUF="${GCC:-}" ; else SUF="" ; fi # randomize keeping suffix on links
+  - export CC="ccache gcc${SUF}" && export CXX="ccache g++${SUF}" && export FC="ccache gfortran${SUF}" && export F77="ccache gfortran${SUF}" || true
+  - echo "\$SUF=${SUF}, \$CC=${CC}, \$CXX=${CXX}, \$FC=${FC} \$F77=${F77}"
+  - if [[ -n "${GCC}" ]]; then ls -l "$(which gcc$GCC)"  && ln -s "$(which gcc$GCC)" "${HOME}/bin/gcc${SUF}" ; fi
+  - if [[ -n "${GCC}" ]]; then ls -l "$(which g++$GCC)"  && ln -s "$(which g++$GCC)" "${HOME}/bin/g++${SUF}" ; fi
+  - if [[ -n "${GCC}" ]]; then ls -l "$(which gfortran$GCC)" && ln -s "$(which gfortran$GCC)" "${HOME}/bin/gfortran${SUF}"; fi
+  - $CC --version && $CXX --version && $FC --version || true # allow missing gfortran
+  - |
+    if [[ "${MPI:-}" == "MPICH" ]]; then
+      cd "${HOME}/mpich" && ls -a \
+      && if [[ ! -f "./mpich-3.2.tar.gz" ]]; then wget --continue --no-clobber 'http://www.mpich.org/static/downloads/3.2/mpich-3.2.tar.gz' ; else echo "Using cached MPICH tarball"; fi \
+      && if [[ ! -d "./mpich-3.2" ]]; then tar -xzvf mpich-3.2.tar.gz --keep-newer-files 2>&1 | tee >(head) | tail || true ; else echo "Using cached MPICH source dir"; fi \
+      && cd build \
+      && if [[ ! -f Makefile ]]; then ../mpich-3.2/configure --prefix="${HOME}/usr" 2>&1 | tee >(head) | tail -500 ; else echo "Using cached MPICH configuration, not reconfiguring"; fi \
+      && make -j 2>&1 | tee >(head) | tail -500 && make install 2>&1 | tee >(head) | tail -500 \
+      && cd "${TRAVIS_BUILD_DIR}" \
+      && which mpicc && mpicc --version && which mpif90 && mpif90 --version || true \
+      && mpicc -show && mpif90 -show || true
+    fi
```

Comment:
Git info about system compilers, enable ccache for building MPICH, and try to do as little work as possible to build MPICH, via hitting the Travis-CI cache. If things change, this may require deleting the travis-ci caches to fix stuff, but I don't foresee that being an issue. Also, just print head and tail -500 for MPICH install stuff... to reduce verbosity and try to prevent needing to fetch raw log from S3.